lib/streamlit/web/server/oidc_mixin.py

```diff
@@ -0,0 +1,102 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tornado.web
+from authlib.integrations.base_client import (  # type: ignore[import-untyped]
+    BaseApp,
+    BaseOAuth,
+    OAuth2Mixin,
+    OAuthError,
+    OpenIDMixin,
+)
+from authlib.integrations.requests_client import (  # type: ignore[import-untyped]
+    OAuth2Session,
+)
+
+from streamlit.web.server.authlib_tornado_integration import TornadoIntegration
+
+
+class TornadoOAuth2App(OAuth2Mixin, OpenIDMixin, BaseApp):  # type: ignore[misc]
+    client_cls = OAuth2Session
+
+    def load_server_metadata(self):
+        result = super().load_server_metadata()
+        if "S256" in result.get("code_challenge_methods_supported", []):
+            # Todo[kajarenc]: maybe not set code_challenge_method to S256 if
+            #  it is explicitly stated in the auth section.
+            self.client_kwargs["code_challenge_method"] = "S256"
+        return result
+
+    def save_authorize_data(self, **kwargs):
+        state = kwargs.pop("state", None)
+        if state:
+            assert self.framework.cache is not None
+            session = None
+            self.framework.set_state_data(session, state, kwargs)
+        else:
+            raise RuntimeError("Missing state value")
+
+    def authorize_redirect(
+        self, request_handler: tornado.web.RequestHandler, redirect_uri=None, **kwargs
+    ):
+        """Create a HTTP Redirect for Authorization Endpoint.
+
+        :param request: HTTP request instance from Starlette view.
+        :param redirect_uri: Callback or redirect URI for authorization.
+        :param kwargs: Extra parameters to include.
+        :return: A HTTP redirect response.
+        """
+        rv = self.create_authorization_url(redirect_uri, **kwargs)
+        self.save_authorize_data(redirect_uri=redirect_uri, **rv)
+        request_handler.redirect(rv["url"], status=302)
+
+    def authorize_access_token(
+        self, request_handler: tornado.web.RequestHandler, **kwargs
+    ):
+        error = request_handler.get_argument("error", None)
+        if error:
+            description = request_handler.get_argument("error_description", None)
+            raise OAuthError(error=error, description=description)
+
+        params = {
+            "code": request_handler.get_argument("code"),
+            "state": request_handler.get_argument("state"),
+        }
+
+        assert self.framework.cache is not None
+        session = None
+
+        claims_options = kwargs.pop("claims_options", None)
+        state_data = self.framework.get_state_data(session, params.get("state"))
+        self.framework.clear_state_data(session, params.get("state"))
+        params = self._format_state_params(state_data, params)
```

Comment:
@kajarenc  After this line I need to add
```# Get redirect_uri from state_data or params
        if 'headers' not in kwargs:
            kwargs['headers'] = {}
        redirect_uri = state_data.get('redirect_uri') or params.get('redirect_uri')
        kwargs['headers']['Origin'] = redirect_uri```

to avoid `authlib.integrations.base_client.errors.OAuthError: invalid_request: AADSTS9002327: Tokens issued for the 'Single-Page Application' client-type may only be redeemed via cross-origin requests.`

Reply:
It would be unsafe, I think specifying correct same `redirect_uri` both `secrets.toml` and Microsoft app should solve the problem.